yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+      try {
+        var stopped = false;
+        while (!stopped) {
+          take match {
+            case StopExecution() =>
+              // stop: exit the loop
+              stopped = true
+            case Start() =>
+              // initial read; may be bigger
+              doRefresh(true)
+            case RefreshRequest(time) =>
+              // requested refresh operation
+              doRefresh(false)
+          }
+          // it is only after processing the
+          // message that the message process counter
+          // is incremented
+          _messagesProcessed.incrementAndGet()
+
+        }
+      } finally {
+        closeQueryClient();
+        running.set(false)
+      }
+    }
+
+    /**
+     * Do the real refresh.
+     * This contains the decision making as when to refresh, which consists of
+     * 1. the refresh interval == 0 (always)
+     * 2. the last refresh was outside the window.
+     *
+     * There isn't a special check for "never updated", as this
+     * would only be inside the window in test cases with a small
+     * simulated clock.
+     *
+     * @param startup a flag to indicate this is the startup retrieval with different window policy
+     */
+    private def doRefresh(startup: Boolean): Unit = {
+      val current = now()
+      if (refreshInterval == 0
+          || ((now() - _lastRefreshAttemptTime.get()) > refreshInterval )) {
+        logDebug("refresh triggered")
+        listAndCacheApplications(true)
+        _lastRefreshAttemptTime.set(now())
+        _refreshesExecuted.incrementAndGet()
+      }
+    }
+
+    /**
+     * Get the next action; increment the [[_messagesProcessed]]
+     * counter after
+     * @return the next action
+     */
+    def take(): Refresher.this.RefreshActions = {
+      val action = queue.take()
+      action
+    }
+
+    /**
+     * Flag to indicate the refresher thread is running
+     * @return
+     */
+    def isRunning(): Boolean  = {
+      running.get()
+    }
+
+    /**
+     * Get the last refresh time
+     * @return the last refresh time
+     */
+    def lastRefreshAttemptTime: Long = {
+      _lastRefreshAttemptTime.get()
+    }
+
+    /**
+     * Get count of messages processed.
+     * This will be at least equal to
+     * the number of refreshes executed
+     * @return
+     */
+    def messagesProcessed: Long = {
+      _messagesProcessed.get()
+    }
+
+    /**
+     * The number of actual refreshes triggered
+     * @return a count of refreshes
+     */
+    def refreshesExecuted: Long = {
+      _refreshesExecuted.get()
+    }
+
+    override def toString: String = {
+      s"Refresher running = $isRunning queue size = ${queue.size()};" +
+        s" processed = $messagesProcessed;" +
+        s" window = $refreshInterval mS;" +
+        s" refreshes executed = ${_refreshesExecuted.get()};" +
+        s" last refresh attempt = " + timeShort(lastRefreshAttemptTime, "never") + ";"
+    }
+  } // end class Refresher
```

Comment:
nit: we don't usually mark the end of things like this. editors are pretty good at doing bracket matching these days.